### PR TITLE
Implemented the pod and service check assertions

### PR DESCRIPTION
**Pull request by sankarpn**
Added assertion implementation for domain pods and services

### CODE REVIEW

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/TestAssertions.java [hunk 1]
```diff
@@ -25,57 +28,84 @@
...
 
+  /**
+   * Check if a Kubernetes pod exists in any state.
+   *
+   * @param podName name of the pod to check for
+   * @param domainUID WebLogic domain uid in which the pod belongs
+   * @param namespace in which the pod exists
+   * @return true if the pod exists in the namespace otherwise false
+   */
+  public static Callable<Boolean> podExists(String podName, String domainUID, String namespace) throws ApiException {
```
Comment (rjeberhard) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/TestAssertions.java:61]: Let's discuss in the meeting.  It seems that these assertions are insufficient (I know that they model exactly what was in the previous tests).  Just checking that the pod exists, is ready, or is terminating leaves off lots of possible errors.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/TestAssertions.java [hunk 2]
```diff
@@ -25,57 +28,84 @@
...
    * Check is a service exists in given namespace.
-   * @param serviceName  the name of the service to check for
+   *
+   * @param serviceName the name of the service to check for
    * @param namespace in which the service is running
    * @return true if the service exists otherwise false
    */
-  public static boolean serviceReady(String serviceName, String namespace) {
-    return Kubernetes.serviceCreated(serviceName, namespace);
+  public static boolean serviceReady(String serviceName, HashMap label, String namespace) throws ApiException {
```
Comment (rjeberhard) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/TestAssertions.java:102]: HashMap -> Map<String, String>

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java [hunk 3]
```diff
@@ -5,17 +5,61 @@
...
+  private static ApiextensionsV1beta1Api apiextensionsV1beta1Api = new ApiextensionsV1beta1Api();
+  private static ApisApi apisApi = new ApisApi();
+
+
+  /**
+   * Check if the Domain CRD exists
+   * @return true if domains.weblogic.oracle CRD exists otherwise false
+   * @throws Exception
+   */
+  public static boolean isCRDExists() throws Exception {
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java:32]: doesCRDExist

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java [hunk 4]
```diff
@@ -5,17 +5,61 @@
...
+    try {
+      V1beta1CustomResourceDefinition domainBetaCrd =
+          apiextensionsV1beta1Api.readCustomResourceDefinition(
+              "domains.weblogic.oracle", null, null, null);
+      assertNotNull(domainBetaCrd);
+      logger.info("domainBetaCrd is not null");
+      return true;
+    } catch (ApiException aex) {
+      if (aex.getCode() == 404) {
+        assertTrue(false, "Expected CRD domains.weblogic.oracle existed.");
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java:42]: the message is unclear - it reads like you are saying that the CRD that you expected to exist DOES exist

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java [hunk 5]
```diff
@@ -5,17 +5,61 @@
...
-      String[] pods = {};
-      for (String pod : pods) {
-        if (!Kubernetes.podRunning(pod, domainUID, namespace).call().booleanValue()) {
-          return false;
-        }
-      }
-      return true;
+      Object domainObject =
+          customObjectsApi.getNamespacedCustomObject(
+              "weblogic.oracle", "v2", namespace, "domains", domainUID);
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java:60]: is v2 right??

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 6]
```diff
@@ -3,30 +3,268 @@
...
-    };
+  /**
+   * Checks if a pod exists in a given namespace in any state.
+   * @param namespace in which to check for the pod existence
+   * @param domainUid the label the pod is decorated with
+   * @param podName name of the pod to check for
+   * @return true if pod exists otherwise false
+   * @throws ApiException when there is error in querying the cluster
+   */
+  public static boolean isPodExists(String namespace, String domainUid, String podName) throws ApiException {
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:54]: dosPodExist

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 7]
```diff
@@ -3,30 +3,268 @@
...
+    logger.info("Checking if the pod exists in namespace");
+    String labelSelector = null;
+    if (domainUid != null) {
+      labelSelector = String.format("weblogic.domainUID in (%s)", domainUid);
+    }
+    V1Pod pod = getPod(namespace, labelSelector, podName);
+    return pod != null;
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:61]: this is too magical. what does this mean?
do your comparison and assign the value to a well-named variable, then return the var

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 8]
```diff
@@ -3,30 +3,268 @@
...
+   * @return V1Pod object if found otherwise null
+   * @throws ApiException when there is error in querying the cluster
+   */
+  public static V1Pod getPod(String namespace, String labelSelector, String podName) throws ApiException {
+    V1PodList v1PodList =
+        coreV1Api.listNamespacedPod(
+            namespace,
+            Boolean.FALSE.toString(),
+            Boolean.FALSE,
+            null,
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:128]: put comments after each value to explain what they are

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 9]
```diff
@@ -3,30 +3,268 @@
...
+
+  /**
+   * Checks if a Kubernetes service object exists in a given namespace.
+   * @param serviceName name of the service to check for
+   * @param label the key value pair with which the service is decorated with
+   * @param namespace the namespace in which to check for the service
+   * @return true if the service is found otherwise false
+   * @throws ApiException when there is error in querying the cluster
+   */
+  public static boolean isServiceCreated(String serviceName, HashMap label, String namespace) throws ApiException {
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:155]: wasServiceCreated or doesServiceExist

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 10]
```diff
@@ -3,30 +3,268 @@
...
+    if (label != null) {
+      String key = label.keySet().iterator().next().toString();
+      String value = label.get(key).toString();
+      labelSelector = String.format("%s in (%s)", key, value);
+      logger.info(labelSelector);
+    }
+    V1ServiceList v1ServiceList
+        = coreV1Api.listServiceForAllNamespaces(
+        Boolean.FALSE,
+        null,
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:178]: explain params

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 11]
```diff
@@ -3,30 +3,268 @@
...
+        labelSelector,
+        null,
+        Boolean.FALSE.toString(),
+        null,
+        null,
+        Boolean.FALSE);
+    for (V1Service service : v1ServiceList.getItems()) {
+      if (service.getMetadata().getName().equals(serviceName.trim())
+          && service.getMetadata().getNamespace().equals(namespace.trim())) {
+        logger.info("Service Name :" + service.getMetadata().getName());
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:189]: put space after colon

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 12]
```diff
@@ -3,30 +3,268 @@
...
+   * @param labelSelectors with which the pods are decorated
+   * @throws ApiException when there is error in querying the cluster
+   */
+  public static void listPods(String namespace, String labelSelectors) throws ApiException {
+    V1PodList v1PodList =
+        coreV1Api.listNamespacedPod(
+            namespace,
+            Boolean.FALSE.toString(),
+            Boolean.FALSE,
+            null,
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:217]: explain

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 13]
```diff
@@ -3,30 +3,268 @@
...
+   * This method can be used as diagnostic tool to get the details of services.
+   * @param namespace in which to list all services
+   * @param labelSelectors  with which the services are decorated
+   * @throws ApiException when there is error in querying the cluster
+   */
+  public static void listServices(String namespace, String labelSelectors) throws ApiException {
+    V1ServiceList v1ServiceList
+        = coreV1Api.listServiceForAllNamespaces(
+        Boolean.FALSE,
+        null,
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:239]: explain

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java [hunk 14]
```diff
@@ -3,30 +3,268 @@
...
+        labelSelectors,
+        null,
+        Boolean.FALSE.toString(),
+        null,
+        null,
+        Boolean.FALSE);
+    List<V1Service> items = v1ServiceList.getItems();
+    logger.info(Arrays.toString(items.toArray()));
+    for (V1Service service : items) {
+      logger.info("Service Name :" + service.getMetadata().getName());
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Kubernetes.java:250]: space after colon

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Operator.java [hunk 15]
```diff
@@ -3,30 +3,36 @@
...
-import java.util.Random;
+import java.util.HashMap;
 import java.util.concurrent.Callable;
 
-import static oracle.weblogic.kubernetes.extensions.LoggedTest.logger;
+import io.kubernetes.client.openapi.ApiException;
+import oracle.weblogic.kubernetes.extensions.LoggedTest;
 
-public class Operator {
+public class Operator implements LoggedTest {
```
Comment (markxnelson) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Operator.java:12]: this is not a test -- should not implement LoggedTest
just use a static import to get the logger (this is done in other assertion impls)

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java (hunk 5 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java:60]: what should be the version value ?

Comment (markxnelson) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java:60]: @rjeberhard is v2 still right?

Comment (markxnelson) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java:60]: i see `v7` i think? 

```apiVersion: weblogic.oracle/v7
kind: Domain
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"weblogic.oracle/v6","kind":"Domain","metadata":{"annotations":{},"labels":{"weblogic.domainUID":"domain1","weblogic.resourceVersion":"domain-v2"},"name":"domain1","namespace":"mark"},"spec":{"adminServer":{"adminService":{"channels":[{"channelName":"default","nodePort":30751}]},"serverStartState":"RUNNING"},"clusters":[{"clusterName":"cluster-1","replicas":2,"serverStartState":"RUNNING"}],"dataHome":"","domainHome":"/u01/oracle/user_projects/domains/domain1","domainHomeInImage":true,"image":"phx.ocir.io/stevengreenberginc/mark/domain-home-in-image:12.2.1.3","imagePullPolicy":"IfNotPresent","includeServerOutInPodLog":true,"serverPod":{"env":[{"name":"JAVA_OPTIONS","value":"-Dweblogic.StdoutDebugEnabled=false"},{"name":"USER_MEM_ARGS","value":"-Djava.security.egd=file:/dev/./urandom "}]},"serverStartPolicy":"IF_NEEDED","webLogicCredentialsSecret":{"name":"domain1-weblogic-credentials"}}}
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/TestAssertions.java [hunk 16]
```diff
@@ -3,19 +3,23 @@
...
+import io.kubernetes.client.openapi.ApiException;
 import oracle.weblogic.kubernetes.assertions.impl.Domain;
 import oracle.weblogic.kubernetes.assertions.impl.Kubernetes;
 import oracle.weblogic.kubernetes.assertions.impl.Operator;
 
 
-// as in the actions, it is intended tests only use these assertaions and do
-// not go direct to the impl classes
+/**
+ * General assertions needed by the tests to validate CRD, Domain, Pods etc., *
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/TestAssertions.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java [hunk 17]
```diff
@@ -5,17 +5,63 @@
...
+  private static CustomObjectsApi customObjectsApi = new CustomObjectsApi();
+  private static ApiextensionsV1Api apiextensionsV1Api = new ApiextensionsV1Api();
+  private static ApiextensionsV1beta1Api apiextensionsV1beta1Api = new ApiextensionsV1beta1Api();
+  private static ApisApi apisApi = new ApisApi();
+
+
+  /**
+   * Check if the Domain CRD exists
+   * @return true if domains.weblogic.oracle CRD exists otherwise false
+   * @throws Exception
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Domain.java]: fixed

**Pull request merged**